Remove the usage of the deprecated distutils module

DavidACrompton opened the pull request:
## Issue description
As Python 3.12 will remove the distutils module per  [PEP 632](https://peps.python.org/pep-0632/), in order to support 3.12 Panda needs to remove distutils in favour of the setuptools module.

Related Issue: #1395 

## Solution description
I have solved this by:

1. Creating a locations.py file for FreezeTool to use, that on Python 3.10 and above uses the sysconfig module to get paths, and uses distutils on lower versions
2. Changed FreezeTool.py, commands.py, and installers.py to use the built in logging module, as opposed to using distutils.log
3. Changed makepanda back to RDB's commit https://github.com/panda3d/panda3d/commit/2fcacd1bab1226ea0571c5d925daf6028a862f96, where he removes distutils in favour of sysconfig

## Notes

Whilst working on this, I came across issue #1230, which was in response to RDB's commit removing distutils in favour of sysconfig. I have tested this on Ubuntu and Debian itself, and had no issues. The sysconfig module now includes the correct paths https://github.com/panda3d/panda3d/issues/1395#issuecomment-1383126253 

## Checklist
I have done my best to ensure that…
* [x] …I have familiarized myself with the CONTRIBUTING.md file
* [x] …this change follows the coding style and design patterns of the codebase
* [x] …I own the intellectual property rights to this code
* [x] …the intent of this change is clearly explained
* [x] …existing uses of the Panda3D API are not broken
* [ ] …the changed code is adequately covered by the test suite, where possible.

rdb commented on direct/src/dist/locations.py:
```diff
@@ -0,0 +1,16 @@
+import sys
+
+if sys.version_info < (3, 10):
```
I am not sure this check should be for 3.10, unless we are sure that no Ubuntu/Debian versions exist with a broken version of Python 3.10.  I suggest we do some digging to find out where/when they fixed it.

DavidACrompton replied:
@rdb, had a dig around for this on Debian Salsa. Looking at the change log it was Debian `python3.10 (3.10.2-3)` on Fri, 25 Feb 2022, when they re-introduced patching sysconfig. This does mean that there may be some people with a broken sysconfig on python3.10, but updating their python via APT should solve it. 
Ultimately, it's a judgement call for you to make. 
https://salsa.debian.org/cpython-team/python3/-/blob/master/debian/changelog#L202

Moguri replied:
Instead of a version check, I would prefer that we just attempt to use `distutils` and fallback to `sysconfig` on an `ImportError`.

rdb replied:
I'm happy with either an ImportError fallback, a version check for for Python 3.12+ or a version check for Python 3.10.3+.

Being certain that we don't break any build configuration weighs more heavily than possibly having a deprecation warning be shown in some configurations.

rdb commented on makepanda/installpanda.py:
```diff
@@ -13,8 +13,7 @@
 
-# DO NOT CHANGE TO sysconfig - see GitHub issue #1230
-from distutils.sysconfig import get_python_lib
+import sysconfig
```
There's a reason the comment is here.  This would break when building on older Ubuntu versions.

Moguri replied:
We might as well use `locations.py` here too.

rdb replied:
I think the difficulty is that we can't import direct modules in makepanda.  It's not immediately obvious to me what the neatest solution is to avoid duplicating that file.

rdb commented on direct/src/dist/locations.py:
```diff
@@ -0,0 +1,16 @@
+    import sysconfig as sysconf
+    PythonIPath = sysconf.get_path("include")
+    PythonVersion = sysconf.get_config_var("LDVERSION") or sysconf.get_python_version()
+    Python = sysconf.get_config_var('prefix')
```
Not a fan of the naming of these variables

Moguri replied:
I don't either. I would prefer this file to just be a super thin wrapper for `distutils.sysconf`, even keeping the original names:

```python
try:
  from distutils.sysconfig import *

 # not sure if this works or if we need to setup explicit aliases
  __all__ = ['get_python_inc', 'get_config_var', ...] 
except ImportError:
  from sysconfig  import *

  __all__ = ['get_config_var', ...]

  def get_python_inc():
    return get_path('include')
```

rdb commented on direct/src/dist/commands.py:
```diff
@@ -580,6 +582,12 @@ def run(self):
+        """ Override the command's built-in announce function as older versions of setuptools does not
+        use logging style levels.
+        """
+        build_apps.log.log(level, msg)
```
Could you explain this?  Does this interact with the setuptools logging?  I would like not to be using two log systems at the same time.

DavidACrompton replied:
Yes, essentially setuptools includes the disttools logger, but they're monkey-patch the some functions in the distutil  version (https://github.com/pypa/setuptools/blob/main/setuptools/logging.py#L27) to use the builtin logging module. In other places they use the logging module but with the log levels divided by 10. (so WARNING would become WARNING/10), I thought it best to override the `announce` function from setuptools, to use the logging module solely in `commands.py`. Happy to change this.

rdb replied:
I see.  That's annoying.  I can't think of a cleaner solution at the moment.

rdb replied:
I think we have to verify that it's still possible to increase the output verbosity from the command-line with the setuptools verbosity flag

rdb reviewed (changes requested):
Thanks for the PR.  Comments inline.

Moguri commented on makepanda/makepackage.py:
```diff
@@ -8,6 +8,7 @@
 import subprocess
 from makepandacore import *
 from installpanda import *
+import sysconfig
```
Prefer `locations.py` over `sysconfig`

Moguri commented on makepanda/makepandacore.py:
```diff
@@ -6,7 +6,7 @@
 
 import configparser
-from distutils import sysconfig # DO NOT CHANGE to sysconfig - see #1230
+import sysconfig
```
Prefer `locations.py` over `sysconfig`

DavidACrompton commented:
I've changed locations.py as per Moguri's recommendation to make it a wrapper around distutils.sysconfig, and created aliases for the functions that differ in the newer sysconfig. 

I've changed the version check to test for 3.12 or above, just in consideration that although sysconfig  now gets patched in mainline debian, theres still other distros that may lag behind the mainline. As rdb noted, a deprecation warning is preferable to breaking some build configurations. 

I also reverted the makepanda changes, if anyone any suggestions on how the best way to handle sysconfig there for 3.12+?

rdb commented on direct/src/dist/locations.py:
```diff
@@ -0,0 +1,18 @@
+else:
+    from sysconfig import *
+
+    __all__ = ['get_python_inc', 'get_config_var', "get_python_version", "PREFIX", "get_python_lib", "get_config_vars"]
```
This is the same in both cases, put this at the top

Also, use consistent quotation marks

rdb reviewed (changes requested):
As for makepanda, I think we don't have much choice but to copy the locations.py file, or integrate the functions into makepandacore.py

Moguri commented:
> As for makepanda, I think we don't have much choice but to copy the locations.py file, or integrate the functions into makepandacore.py; since makepanda is on death row I don't mind not being super clean about that

Yeah, I am fine with just duplicating the file. It shouldn't need much, if any, updating, and, as you said, makepanda is on the way out anyways.

Moguri commented:
Any updates on this? Now that Python 3.12 is officially out, we should really get this sorted.

rdb commented:
Work continues in #1549.
